### PR TITLE
Razor TagHelperDescriptor Hashing

### PR DESCRIPTION
**Pull request by TanayParikh**
1/2 PRs for https://github.com/dotnet/aspnetcore/issues/23170

- I was running into quite a few hash-collisions so decided to revamp the hashing logic throughout the `TagHelperDescriptor` to minimize collisions. 
- Added (collision) testing.
- Added a `TagHelperDescriptor` memory cache.

The other PR utilizes these changes: https://github.com/dotnet/aspnetcore-tooling/pull/2307

### CODE REVIEW

File: src/Shared/RazorShared/TagHelperDescriptorJsonConverter.cs [hunk 1]
```diff
@@ -5,7 +5,6 @@
 using System.Collections.Generic;
 using Microsoft.AspNetCore.Razor.Language;
 using Newtonsoft.Json;
-using Newtonsoft.Json.Linq;
```
Comment (TanayParikh) [inline src/Shared/RazorShared/TagHelperDescriptorJsonConverter.cs:8]: Shared file update from [tooling](https://github.com/dotnet/aspnetcore-tooling).

File: src/Shared/RazorShared/MemoryCache.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    // We've created our own MemoryCache here, ideally we would use the one in Microsoft.Extensions.Caching.Memory,
+    // but until we update O# that causes an Assembly load problem.
+    internal class MemoryCache<TKey, TValue>
+    {
+        private static readonly int SizeLimit = 1500;
```
Comment (TanayParikh) [inline src/Shared/RazorShared/MemoryCache.cs:15]: Large cache size of 1500 `TagHelperDescriptors` done to ensure it works well for large projects (where these changes are most beneficial).

This should be at most around 3MB:

224 serialized taghelpers are 390KB => 390/224*1500/1024 = 2.55 + CacheEntry overhead ~= 3MB.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorCache.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    public static class TagHelperDescriptorCache
```
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorCache.cs:6]:
```suggestion
    internal static class TagHelperDescriptorCache
```

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorCache.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        private static readonly MemoryCache<int, TagHelperDescriptor> CachedTagHelperDescriptors = new MemoryCache<int, TagHelperDescriptor>();
+
+        // Disable cache for testing purposes
+        internal static bool CacheEnabled { private get; set; } = true;
```
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorCache.cs:11]: Lets make it extra clear and  rename  this `CacheEnabledForTest`

File: src/Shared/RazorShared/MemoryCache.cs (hunk 2 above)
Comment (NTaylorMullen) [reply src/Shared/RazorShared/MemoryCache.cs:15]: Given the  reasoning for the size i s based on TagHelpers and you've  made this memory  cache  configurable for  more then just TagHelpers  i'd recommend taking  this  size limit in through the constructor and having  the TagHelperDescriptorCache set it to 1500

**Review by NTaylorMullen — APPROVED** [review]

File: src/Shared/RazorShared/MemoryCache.cs (hunk 2 above)
Comment (TanayParikh) [reply src/Shared/RazorShared/MemoryCache.cs:15]: I was thinking of making this TagHelper specific through the following changes:

- SizeLimit of 1500 to start
- At 1500 perform `Compact` _but_ only compact entries which haven't been used in the past ~15 seconds (we already store the `LastAccess` time. If all 1500 entries have been used recently, then expand the size by 500. 
- The purpose of this is we don't have a significant perf drop off at 1500 entries.

Would you say this additional complexity is worthwhile or too much of an edge case?

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorCache.cs (hunk 3 above)
Comment (NTaylorMullen) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorCache.cs:6]: Actually, why have this in the language  repo at  all? Seems like  more of a tooling concern given I don't see t his  used  in this PR. Ditto witht he memory cache impl

**Review by NTaylorMullen — CHANGES_REQUESTED** [review]

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorCache.cs (hunk 3 above)
Comment (TanayParikh) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorCache.cs:6]: Won't that make it package internal to `Microsoft.AspNetCore.Razor.Language` and inaccessible in aspnetcore-tooling based namespaces?

**Review by NTaylorMullen — APPROVED** [review]

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorComparer.cs [hunk 5]
```diff
@@ -120,8 +120,50 @@ public virtual int GetHashCode(TagHelperDescriptor descriptor)
...
             hash.Add(descriptor.AssemblyName, StringComparer.Ordinal);
             hash.Add(descriptor.Name, StringComparer.Ordinal);
+            hash.Add(descriptor.DisplayName, StringComparer.Ordinal);
+            hash.Add(descriptor.CaseSensitive ? 1 : 0);
+
+            if (descriptor.BoundAttributes != null)
+            {
+                for (var i = 0; i < descriptor.BoundAttributes.Count; i++)
+                {
+                    hash.Add(descriptor.BoundAttributes[i]);
```
Comment (ajaybhargavb) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorComparer.cs:130]: Would it be better to call `GetHashCode()` on each of the bound attributes? Similiarly for other places. I don't know if it makes a difference. Just wondering.

Comment (TanayParikh) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/TagHelperDescriptorComparer.cs:130]: The `HashCodeCombiner.Add` calls `GetHashCode()` underneath (with appropriate null checks and handling).

**Pull request merged**